File: go/fluentcards/teams_features_test.go
package fluentcards

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// ── Card-level msteams (TeamsCardPropertiesBuilder) ──────────────────────────

func TestTeamsCardFullWidth(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *TextBlockBuilder) { tb.WithText("Hello") }).
		WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) { tc.WithFullWidth() }).
		Build()
	msteams := card["msteams"].(map[string]any)
	assert.Equal(t, "Full", msteams["width"])
}

func TestTeamsCardWithMention(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *TextBlockBuilder) { tb.WithText("Hi <at>John</at>") }).
		WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) { tc.AddMention("John", "user-123") }).
		Build()
	msteams := card["msteams"].(map[string]any)
	entities := msteams["entities"].([]any)
	require.Len(t, entities, 1)
	entity := entities[0].(map[string]any)
	assert.Equal(t, "mention", entity["type"])
	assert.Equal(t, "<at>John</at>", entity["text"])
	mentioned := entity["mentioned"].(map[string]any)
	assert.Equal(t, "user-123", mentioned["id"])
	assert.Equal(t, "John", mentioned["name"])
}

func TestTeamsCardFullWidthWithMentions(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *TextBlockBuilder) { tb.WithText("Hi <at>Alice</at> and <at>Bob</at>") }).
		WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) {
			tc.WithFullWidth().AddMention("Alice", "user-1").AddMention("Bob", "user-2")
		}).
		Build()
	msteams := card["msteams"].(map[string]any)
	assert.Equal(t, "Full", msteams["width"])
	entities := msteams["entities"].([]any)
	assert.Len(t, entities, 2)
}

func TestTeamsCardRaw(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *TextBlockBuilder) { tb.WithText("Hi") }).
		WithTeamsCardRaw(map[string]any{"width": "Full", "custom": true}).
		Build()
	msteams := card["msteams"].(map[string]any)
	assert.Equal(t, "Full", msteams["width"])
	assert.Equal(t, true, msteams["custom"])
}

func TestTeamsCardTypedThenRawConflict(t *testing.T) {
	assert.PanicsWithValue(t,
		"AdaptiveCardBuilder: cannot use both WithTeamsCard and WithTeamsCardRaw on the same card",
		func() {
			NewAdaptiveCardBuilder().
				WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) { tc.WithFullWidth() }).
				WithTeamsCardRaw(map[string]any{"width": "Full"})
		})
}

func TestTeamsCardRawThenTypedConflict(t *testing.T) {
	assert.PanicsWithValue(t,
		"AdaptiveCardBuilder: cannot use both WithTeamsCard and WithTeamsCardRaw on the same card",
		func() {
			NewAdaptiveCardBuilder().
				WithTeamsCardRaw(map[string]any{"width": "Full"}).
				WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) { tc.WithFullWidth() })
		})
}

// ── Action-level msteams (Submit feedback) ───────────────────────────────────

func TestSubmitWithFeedbackHidden(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddActionSet(func(as *ActionSetBuilder) {
			as.AddAction(func(a *ActionBuilder) {
				a.Submit("Click me").WithTeamsSubmitFeedback(func(f *TeamsSubmitPropertiesBuilder) {
					f.WithFeedbackHidden()
				})
			})
		}).
		Build()
	body := card["body"].([]any)
	actionSet := body[0].(map[string]any)
	actions := actionSet["actions"].([]any)
	action := actions[0].(map[string]any)
	msteams := action["msteams"].(map[string]any)
	feedback := msteams["feedback"].(map[string]any)
	assert.Equal(t, true, feedback["hide"])
}

func TestSubmitWithTeamsSubmitRaw(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddActionSet(func(as *ActionSetBuilder) {
			as.AddAction(func(a *ActionBuilder) {
				a.Submit("Click me").WithTeamsSubmitRaw(map[string]any{
					"feedback": map[string]any{"hide": true},
				})
			})
		}).
		Build()
	body := card["body"].([]any)
	action := body[0].(map[string]any)["actions"].([]any)[0].(map[string]any)
	msteams := action["msteams"].(map[string]any)
	feedback := msteams["feedback"].(map[string]any)
	assert.Equal(t, true, feedback["hide"])
}

func TestTeamsSubmitTypedThenRawConflict(t *testing.T) {
	assert.Panics(t, func() {
		NewAdaptiveCardBuilder().
			AddActionSet(func(as *ActionSetBuilder) {
				as.AddAction(func(a *ActionBuilder) {
					a.Submit("Click me").
						WithTeamsSubmitFeedback(func(f *TeamsSubmitPropertiesBuilder) { f.WithFeedbackHidden() }).
						WithTeamsSubmitRaw(map[string]any{"feedback": map[string]any{"hide": true}})
				})
			}).Build()
	})
}

func TestTeamsSubmitRawThenTypedConflict(t *testing.T) {
	assert.Panics(t, func() {
		NewAdaptiveCardBuilder().
			AddActionSet(func(as *ActionSetBuilder) {
				as.AddAction(func(a *ActionBuilder) {
					a.Submit("Click me").
						WithTeamsSubmitRaw(map[string]any{"feedback": map[string]any{"hide": true}}).
						WithTeamsSubmitFeedback(func(f *TeamsSubmitPropertiesBuilder) { f.WithFeedbackHidden() })
				})
			}).Build()
	})
}

// ── Data-level Teams (task/fetch, msteams in data) ───────────────────────────

func TestTeamsTaskFetchShorthand(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddActionSet(func(as *ActionSetBuilder) {
			as.AddAction(func(a *ActionBuilder) {
				a.Submit("Open").WithTeamsTaskFetch()
			})
		}).
		Build()
	body := card["body"].([]any)
	action := body[0].(map[string]any)["actions"].([]any)[0].(map[string]any)
	data := action["data"].(map[string]any)
	msteams := data["msteams"].(map[string]any)
	assert.Equal(t, "task/fetch", msteams["type"])
}

func TestTeamsDataWithCustomProperties(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddActionSet(func(as *ActionSetBuilder) {
			as.AddAction(func(a *ActionBuilder) {
				a.Submit("Open").WithTeamsData(func(d *TeamsDataBuilder) {
					d.WithTaskFetch().WithProperty("customKey", "customValue").WithProperty("count", 42)
				})
			})
		}).
		Build()
	body := card["body"].([]any)
	action := body[0].(map[string]any)["actions"].([]any)[0].(map[string]any)
	data := action["data"].(map[string]any)
	assert.Equal(t, map[string]any{"type": "task/fetch"}, data["msteams"])
	assert.Equal(t, "customValue", data["customKey"])
	assert.Equal(t, 42, data["count"])
}

func TestTeamsDataWithCustomMsteams(t *testing.T) {
	b := newTeamsDataBuilder()
	b.WithMsteams(map[string]any{"type": "task/submit", "custom": true})
	b.WithProperty("key", "value")
	result := b.Build()
	assert.Equal(t, map[string]any{"type": "task/submit", "custom": true}, result["msteams"])
	assert.Equal(t, "value", result["key"])
}

func TestTeamsDataMsteamsPropertyRejected(t *testing.T) {
	assert.Panics(t, func() {
		b := newTeamsDataBuilder()
		b.WithProperty("msteams", map[string]any{"type": "task/fetch"})
	})
}

func TestTeamsDataMsteamsPropertyCaseInsensitive(t *testing.T) {
	assert.Panics(t, func() {
		b := newTeamsDataBuilder()
		b.WithProperty("MSTEAMS", "value")
	})
}

// ── Submit-only gating ───────────────────────────────────────────────────────

func TestTeamsTaskFetchOnExecutePanics(t *testing.T) {
	assert.Panics(t, func() {
		NewAdaptiveCardBuilder().
			AddActionSet(func(as *ActionSetBuilder) {
				as.AddAction(func(a *ActionBuilder) {
					a.Execute("Run").WithTeamsTaskFetch()
				})
			}).Build()
	})
}

func TestTeamsDataOnOpenURLPanics(t *testing.T) {
	assert.Panics(t, func() {
		NewAdaptiveCardBuilder().
			AddActionSet(func(as *ActionSetBuilder) {
				as.AddAction(func(a *ActionBuilder) {
					a.OpenURL("https://example.com").WithTeamsData(func(d *TeamsDataBuilder) { d.WithTaskFetch() })
				})
			}).Build()
	})
}

func TestTeamsSubmitFeedbackOnTogglePanics(t *testing.T) {
	assert.Panics(t, func() {
		NewAdaptiveCardBuilder().
			AddActionSet(func(as *ActionSetBuilder) {
				as.AddAction(func(a *ActionBuilder) {
					a.ToggleVisibility("Toggle").WithTeamsSubmitFeedback(func(f *TeamsSubmitPropertiesBuilder) { f.WithFeedbackHidden() })
				})
			}).Build()
	})
}

func TestTeamsSubmitRawOnExecutePanics(t *testing.T) {
	assert.Panics(t, func() {
		NewAdaptiveCardBuilder().
			AddActionSet(func(as *ActionSetBuilder) {
				as.AddAction(func(a *ActionBuilder) {
					a.Execute("Run").WithTeamsSubmitRaw(map[string]any{"feedback": map[string]any{"hide": true}})
				})
			}).Build()
	})
}

// ── Data conflict detection ──────────────────────────────────────────────────

func TestWithDataThenTeamsDataConflict(t *testing.T) {
	assert.Panics(t, func() {
		NewAdaptiveCardBuilder().
			AddActionSet(func(as *ActionSetBuilder) {
				as.AddAction(func(a *ActionBuilder) {
					a.Submit("Click").WithData(map[string]any{"key": "value"}).WithTeamsData(func(d *TeamsDataBuilder) { d.WithTaskFetch() })
				})
			}).Build()
	})
}

func TestTeamsDataThenWithDataConflict(t *testing.T) {
	assert.Panics(t, func() {
		NewAdaptiveCardBuilder().
			AddActionSet(func(as *ActionSetBuilder) {
				as.AddAction(func(a *ActionBuilder) {
					a.Submit("Click").WithTeamsData(func(d *TeamsDataBuilder) { d.WithTaskFetch() }).WithData(map[string]any{"key": "value"})
				})
			}).Build()
	})
}

// ── Mention validation ───────────────────────────────────────────────────────

func TestMatchedMentionsPass(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *TextBlockBuilder) { tb.WithText("Hi <at>John</at>") }).
		WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) { tc.AddMention("John", "user-123") }).
		Build()
	issues := Validate(card)
	for _, i := range issues {
		assert.NotContains(t, i.Code, "MENTION")
		assert.NotContains(t, i.Code, "AT_TOKEN")
	}
}

func TestOrphanedMentionEntity(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *TextBlockBuilder) { tb.WithText("Hello world") }).
		WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) { tc.AddMention("John", "user-123") }).
		Build()
	issues := Validate(card)
	var orphaned []ValidationIssue
	for _, i := range issues {
		if i.Code == "ORPHANED_MENTION_ENTITY" {
			orphaned = append(orphaned, i)
		}
	}
	require.Len(t, orphaned, 1)
	assert.Equal(t, ValidationSeverityWarning, orphaned[0].Severity)
}

func TestOrphanedAtToken(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *TextBlockBuilder) { tb.WithText("Hi <at>Unknown</at>") }).
		WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) { tc.AddMention("John", "user-123") }).
		Build()
	issues := Validate(card)
	var atIssues []ValidationIssue
	for _, i := range issues {
		if i.Code == "ORPHANED_AT_TOKEN" {
			atIssues = append(atIssues, i)
		}
	}
	require.Len(t, atIssues, 1)
}

// ── Serialization round-trip ─────────────────────────────────────────────────

func TestFullTeamsCardRoundTrip(t *testing.T) {
	card := NewAdaptiveCardBuilder().
		AddTextBlock(func(tb *TextBlockBuilder) { tb.WithText("Hi <at>Alice</at>") }).
		WithTeamsCard(func(tc *TeamsCardPropertiesBuilder) {
			tc.WithFullWidth().AddMention("Alice", "user-1")
		}).
		AddActionSet(func(as *ActionSetBuilder) {
			as.AddAction(func(a *ActionBuilder) {
				a.Submit("Submit").
					WithTeamsTaskFetch().
					WithTeamsSubmitFeedback(func(f *TeamsSubmitPropertiesBuilder) { f.WithFeedbackHidden() })
			})
		}).
		Build()

	jsonStr, err := ToJSON(card)
	require.NoError(t, err)

	var parsed map[string]any
	err = json.Unmarshal([]byte(jsonStr), &parsed)
	require.NoError(t, err)

	msteams := parsed["msteams"].(map[string]any)
	assert.Equal(t, "Full", msteams["width"])
	entities := msteams["entities"].([]any)
	assert.Len(t, entities, 1)
}

func TestFromJSONPreservesMsteams(t *testing.T) {
	original := `{"type":"AdaptiveCard","version":"1.5","$schema":"http://adaptivecards.io/schemas/adaptive-card.json","msteams":{"width":"Full"},"body":[{"type":"TextBlock","text":"Hello"}]}`
	card := FromJSON(original)
	require.NotNil(t, card)
	msteams := card["msteams"].(map[string]any)
	assert.Equal(t, "Full", msteams["width"])
}

File: node/packages/fluent-cards/tests/teams-features.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  AdaptiveCardBuilder,
  ActionBuilder,
  TeamsDataBuilder,
  toJson,
  validate,
  ValidationSeverity,
} from 'fluent-cards';

describe('Teams Features', () => {
  // ── Model & Serialization Tests ──

  describe('Serialization', () => {
    it('serializes width Full', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.withFullWidth())
        .addTextBlock(tb => tb.withText('Hello'))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.msteams.width, 'Full');
    });

    it('omits width when not set', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.addMention('John', '29:123'))
        .addTextBlock(tb => tb.withText('<at>John</at>'))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.msteams.width, undefined);
    });

    it('omits msteams when not set', () => {
      const card = AdaptiveCardBuilder.create()
        .addTextBlock(tb => tb.withText('Hello'))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.msteams, undefined);
    });

    it('serializes mentions', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.addMention('John Doe', '29:1241241'))
        .addTextBlock(tb => tb.withText('Hello <at>John Doe</at>!'))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.msteams.entities.length, 1);
      assert.equal(json.msteams.entities[0].type, 'mention');
      assert.equal(json.msteams.entities[0].text, '<at>John Doe</at>');
      assert.equal(json.msteams.entities[0].mentioned.id, '29:1241241');
      assert.equal(json.msteams.entities[0].mentioned.name, 'John Doe');
    });

    it('serializes submit feedback', () => {
      const card = AdaptiveCardBuilder.create()
        .addTextBlock(tb => tb.withText('Test'))
        .addAction(a => a
          .submit('Silent Submit')
          .withTeamsSubmitFeedback(ts => ts.withFeedbackHidden()))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.actions[0].msteams.feedback.hide, true);
    });

    it('serializes msteams and custom properties in data', () => {
      const card = AdaptiveCardBuilder.create()
        .addTextBlock(tb => tb.withText('Test'))
        .addAction(a => a
          .submit('Task Module')
          .withTeamsData(td => td
            .withTaskFetch()
            .withProperty('customField', 'customValue')))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.actions[0].data.msteams.type, 'task/fetch');
      assert.equal(json.actions[0].data.customField, 'customValue');
    });

    it('serializes msteams only in data', () => {
      const card = AdaptiveCardBuilder.create()
        .addTextBlock(tb => tb.withText('Test'))
        .addAction(a => a.submit('Module').withTeamsTaskFetch())
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.actions[0].data.msteams.type, 'task/fetch');
      assert.deepEqual(Object.keys(json.actions[0].data), ['msteams']);
    });

    it('serializes custom properties only in data', () => {
      const card = AdaptiveCardBuilder.create()
        .addTextBlock(tb => tb.withText('Test'))
        .addAction(a => a
          .submit('Feedback')
          .withTeamsData(td => td
            .withProperty('message', 'Clicked')
            .withProperty('source', 'card')))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.actions[0].data.message, 'Clicked');
      assert.equal(json.actions[0].data.source, 'card');
      assert.equal(json.actions[0].data.msteams, undefined);
    });
  });

  // ── Builder Tests ──

  describe('Builders', () => {
    it('withTeamsCard sets width', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.withFullWidth())
        .addTextBlock(tb => tb.withText('Test'))
        .build();
      assert.equal(card.msteams?.width, 'Full');
    });

    it('addMention auto-generates at text', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.addMention('John', '29:123'))
        .addTextBlock(tb => tb.withText('<at>John</at>'))
        .build();
      assert.equal(card.msteams?.entities?.length, 1);
      assert.equal(card.msteams!.entities![0].text, '<at>John</at>');
      assert.equal(card.msteams!.entities![0].mentioned.id, '29:123');
    });

    it('supports multiple mentions', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.addMention('Alice', '29:111').addMention('Bob', '29:222'))
        .addTextBlock(tb => tb.withText('<at>Alice</at> and <at>Bob</at>'))
        .build();
      assert.equal(card.msteams!.entities!.length, 2);
    });

    it('withTeamsData with all property types', () => {
      const b = new TeamsDataBuilder();
      b.withProperty('str', 'hello');
      b.withProperty('num', 42);
      b.withProperty('bool', true);
      b.withProperty('obj', { nested: 'value' });
      const result = b.build();
      assert.equal(result.str, 'hello');
      assert.equal(result.num, 42);
      assert.equal(result.bool, true);
      assert.deepEqual(result.obj, { nested: 'value' });
    });
  });

  // ── Submit-Only Gating Tests ──

  describe('Submit-only gating', () => {
    it('withTeamsTaskFetch on execute throws', () => {
      const b = new ActionBuilder();
      b.execute('Test');
      assert.throws(() => b.withTeamsTaskFetch(), /Submit/);
    });

    it('withTeamsData on execute throws', () => {
      const b = new ActionBuilder();
      b.execute('Test');
      assert.throws(() => b.withTeamsData(td => td.withTaskFetch()), /Submit/);
    });

    it('withTeamsSubmitFeedback on execute throws', () => {
      const b = new ActionBuilder();
      b.execute('Test');
      assert.throws(() => b.withTeamsSubmitFeedback(ts => ts.withFeedbackHidden()), /Submit/);
    });

    it('withTeamsTaskFetch on openUrl throws', () => {
      const b = new ActionBuilder();
      b.openUrl('https://example.com');
      assert.throws(() => b.withTeamsTaskFetch(), /Submit/);
    });
  });

  // ── Conflict Detection Tests ──

  describe('Conflict detection', () => {
    it('withTeamsData then withData throws', () => {
      const b = new ActionBuilder();
      b.submit('Test');
      b.withTeamsData(td => td.withTaskFetch());
      assert.throws(() => b.withData({ key: 'value' }), /withData/);
    });

    it('withData then withTeamsTaskFetch throws', () => {
      const b = new ActionBuilder();
      b.submit('Test');
      b.withData({ key: 'value' });
      assert.throws(() => b.withTeamsTaskFetch(), /withData/);
    });

    it('withTeamsCard then withTeamsCardRaw throws', () => {
      const b = AdaptiveCardBuilder.create();
      b.withTeamsCard(t => t.withFullWidth());
      assert.throws(() => b.withTeamsCardRaw({ width: 'Full' }), /withTeamsCard/);
    });

    it('withTeamsCardRaw then withTeamsCard throws', () => {
      const b = AdaptiveCardBuilder.create();
      b.withTeamsCardRaw({ width: 'Full' });
      assert.throws(() => b.withTeamsCard(t => t.withFullWidth()), /withTeamsCard/);
    });

    it('withTeamsSubmitFeedback then withTeamsSubmitRaw throws', () => {
      const b = new ActionBuilder();
      b.submit('Test');
      b.withTeamsSubmitFeedback(ts => ts.withFeedbackHidden());
      assert.throws(() => b.withTeamsSubmitRaw({ feedback: { hide: true } }), /withTeamsSubmitFeedback/);
    });

    it('TeamsDataBuilder rejects msteams key', () => {
      const b = new TeamsDataBuilder();
      assert.throws(() => b.withProperty('msteams', 'value'), /msteams/);
    });

    it('TeamsDataBuilder rejects msteams key case-insensitive', () => {
      const b = new TeamsDataBuilder();
      assert.throws(() => b.withProperty('MsTeams', 'value'), /msteams/);
    });
  });

  // ── Validation Tests ──

  describe('Validation', () => {
    it('warns on orphaned mention entity', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.addMention('John', '29:123'))
        .addTextBlock(tb => tb.withText('Hello world'))
        .build();
      const issues = validate(card);
      assert.ok(issues.some(i => i.code === 'ORPHANED_MENTION_ENTITY'));
    });

    it('warns on orphaned at-token', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.addMention('Alice', '29:456'))
        .addTextBlock(tb => tb.withText('Hello <at>John</at>'))
        .build();
      const issues = validate(card);
      assert.ok(issues.some(i => i.code === 'ORPHANED_AT_TOKEN'));
    });

    it('passes matched mentions', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCard(t => t.addMention('John', '29:123').addMention('Alice', '29:456'))
        .addTextBlock(tb => tb.withText('Hello <at>John</at> and <at>Alice</at>'))
        .build();
      const issues = validate(card);
      assert.ok(!issues.some(i => i.code === 'ORPHANED_MENTION_ENTITY'));
      assert.ok(!issues.some(i => i.code === 'ORPHANED_AT_TOKEN'));
    });
  });

  // ── Integration Tests ──

  describe('Integration', () => {
    it('full Teams card matches expected shape', () => {
      const card = AdaptiveCardBuilder.create()
        .withVersion('1.5')
        .withTeamsCard(teams => teams
          .withFullWidth()
          .addMention('John Doe', '29:1241241'))
        .addTextBlock(t => t.withText('Hello <at>John Doe</at>!'))
        .addAction(a => a.submit('Open Task Module').withTeamsTaskFetch())
        .addAction(a => a.submit('Silent').withTeamsSubmitFeedback(ts => ts.withFeedbackHidden()))
        .build();

      const json = JSON.parse(toJson(card));
      assert.equal(json.msteams.width, 'Full');
      assert.equal(json.msteams.entities.length, 1);
      assert.equal(json.actions[0].data.msteams.type, 'task/fetch');
      assert.equal(json.actions[1].msteams.feedback.hide, true);
    });

    it('existing cards unaffected', () => {
      const card = AdaptiveCardBuilder.create()
        .addTextBlock(tb => tb.withText('Hello'))
        .addAction(a => a.submit('OK').withData({ key: 'value' }))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.msteams, undefined);
      assert.equal(json.actions[0].type, 'Action.Submit');
    });

    it('withTeamsCardRaw passes through', () => {
      const card = AdaptiveCardBuilder.create()
        .withTeamsCardRaw({ width: 'Full' })
        .addTextBlock(tb => tb.withText('Test'))
        .build();
      assert.equal(card.msteams?.width, 'Full');
    });

    it('withTeamsSubmitRaw serializes correctly', () => {
      const card = AdaptiveCardBuilder.create()
        .addTextBlock(tb => tb.withText('Test'))
        .addAction(a => a
          .submit('Silent')
          .withTeamsSubmitRaw({ feedback: { hide: true } }))
        .build();
      const json = JSON.parse(toJson(card));
      assert.equal(json.actions[0].msteams.feedback.hide, true);
    });
  });
});
